lib/connectors/memory.js

```diff
@@ -207,18 +207,27 @@ Memory.prototype.create = function create(model, data, options, callback) {
   var idName = this.idName(model);
   id = (props[idName] && props[idName].type && props[idName].type(id)) || id;
   this.setIdValue(model, data, id);
-  if(!this.collection(model)) {
+  if (!this.collection(model)) {
     this.collection(model, {});
   }
 
-  if (this.collection(model)[id]) {
-    return process.nextTick(function() {
-      callback(new Error('Duplicate entry for ' + model + '.' + idName));
-    });
-  }
+  if (this.collection(model)[id])
+    return callback(new Error('Duplicate entry for ' + model + '.' + idName));
```

Comment:
Hey Raymond,

This is the reason why we decided to have this refactoring for create and use synchronous _create:

Miroslav: "it may be better to not use `this.create`, because it's not clear when exactly it creates the record (whether it guarantees to create it in the same turn of even loop)
and do what Ryan suggested - refactor the creation into a private method that updates the collection in a sync way (no file writing!) and then call it from both `this.create` and `this.findOrCreate"

>It's very confusing to see a synchronous function that takes a callback. Why do we need to make _create synchronous?

I was returning or throwing an exception for this function but I understood that is not a good practice based on [this comment](https://github.com/strongloop/loopback-datasource-juggler/pull/815#discussion_r48939890)